### PR TITLE
[content-visibility] Improve skipped content information in RenderStyle

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 682cdce8ad3fcb142ee4fbb6709c7706d0b2f142
<pre>
[content-visibility] Improve skipped content information in RenderStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=259042">https://bugs.webkit.org/show_bug.cgi?id=259042</a>

Reviewed by Simon Fraser.

Improve skipped content information in RenderStyle because beside the fact that
something is skipped content we also want to know if the skipped content is in
a hidden or auto content-visibility subtree.

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-087-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-087.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasFocusableStyle const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContent const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::effectivePointerEvents const):
(WebCore::RenderStyle::skippedContentReason const):
(WebCore::RenderStyle::effectiveSkippedContent const): Deleted.
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setEffectiveInert):
(WebCore::RenderStyle::setSkippedContentReason):
(WebCore::RenderStyle::setEffectiveSkippedContent): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateRenderTree):
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const): c-v: hidden overrides auto as relevancy reason

Canonical link: <a href="https://commits.webkit.org/267128@main">https://commits.webkit.org/267128@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/8a070893997513ca540b41463fba0f1593d45656

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15771 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14801 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15956 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19087 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16185 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16432 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13427 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18287 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13680 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14239 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14680 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/14404 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17665 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15000 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12715 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14247 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14079 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3761 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18616 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14822 "Built successfully") | | |

### CODE REVIEW

**Comment by nt1m** [comment]
> Improve skipped content information in RenderStyle because beside
the fact that something is skipped content we also want to know
if the skipped content is in a hidden or auto content-visibility
subtree.

I'm curious why we need to know this. Isn't content-visibility: auto fully working off the DOM tree?

**Comment by rwlbuis** [comment]
> > Improve skipped content information in RenderStyle because beside
> > the fact that something is skipped content we also want to know
> > if the skipped content is in a hidden or auto content-visibility
> > subtree.
> 
> I'm curious why we need to know this. Isn't content-visibility: auto fully working off the DOM tree?

I found at least one case where this information makes things optimal. When testing for isFocusable, we need to check for skipped content in c-v: hidden subtrees, but not auto subtrees. It is possible to do this by walking the ancestors, but it is more efficient to just ask the RenderStyle directly.
I only noticed this since hooking things up with updateRendering, so it is possible the different timing exposes this only now.

File: Source/WebCore/rendering/style/RenderStyle.h [hunk 1]
```diff
@@ -694,7 +694,7 @@ class RenderStyle {
 
     inline ContentVisibility contentVisibility() const;
 
-    inline bool effectiveSkippedContent() const;
+    inline ContentVisibility effectiveSkippedContent() const;
```
Comment (nt1m) [inline Source/WebCore/rendering/style/RenderStyle.h]: Can we have this be something like:

```
inline std::optional<ContentVisibility> skippedContentReason() const;
```

**Comment by rwlbuis** [comment]
Note, this still needs a test for focusing in a nested c-v: hidden/auto subtree.

File: Source/WebCore/rendering/style/RenderStyle.h (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/style/RenderStyle.h]: Done.

**Comment by rwlbuis** [comment]
@ntim can you have a look? Note that renaming effectiveSkippedContent is still difficult for me since the concept is a bit non standard so is is not easy to name properly.

**Comment by rwlbuis** [comment]
@smfr PTAL, and feel free to do some a name suggestion for this flag, as usual naming is the hardest part :)

File: Source/WebCore/rendering/style/StyleRareInheritedData.h [hunk 2]
```diff
@@ -165,7 +165,7 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
 
     unsigned isInSubtreeWithBlendMode : 1;
 
-    unsigned effectiveSkippedContent : 1;
+    unsigned effectiveSkippedContent : 2;
```
Comment (smfr) [inline Source/WebCore/rendering/style/StyleRareInheritedData.h]: Please add a comment with the name of the enum store here.

**Review by smfr — APPROVED** [review]

**Pull request merged**